### PR TITLE
StandaloneMmPkg/MmIpl: Correct unblocked memory regions attribute

### PR DESCRIPTION
**Pull request by hongbin123**
When PcdCpuSmmRestrictedMemoryAccess is FALSE, resource HOB should exclude unblocked memory regions.

- [ ] Breaking change?
  - **Breaking change** - Does this PR cause a break in build or boot behavior?
  - Examples: Does it add a new library class or move a module to a different repo.
- [ ] Impacts security?
  - **Security** - Does this PR have a direct security impact?
  - Examples: Crypto algorithm change or buffer overflow fix.
- [ ] Includes tests?
  - **Tests** - Does this PR include any explicit test code?
  - Examples: Unit tests or integration tests.

### CODE REVIEW

**Review by jiaxinwu — APPROVED** [review]

File: StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c [hunk 1]
```diff
@@ -450,6 +450,66 @@ MmIplBuildResourceHobForUnblockedRegion (
...
+
+  Index        = 0;
+  MemoryRegion = NULL;
+
+  //
+  // Get unblock memory ranges count
+  //
+  GuidHob = GetFirstGuidHob (&gMmUnblockRegionHobGuid);
+  while (GuidHob != NULL) {
+    UnblockRegion = GET_GUID_HOB_DATA (GuidHob);
```
Comment (niruiyu) [inline StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c:479]: Unnecessary code.

Comment (hongbin123) [reply StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c:479]: ok, will remove

File: StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c [hunk 2]
```diff
@@ -843,23 +916,33 @@ CreateMmFoundationHobList (
...
     //
     // All system memory (DRAM) is accessible.
     // When SMM Profile is enabled:
     //   * Access to regions reported from MmPlatformHobProducerLib do not require logging.
     //   * Access to other system memory requires logging.
     //
-    MmIplBuildResourceHobForAllSystemMemory (FoundationHobList + UsedSize, &HobLength, PlatformHobList, PlatformHobSize, Block, MmProfileDataHob);
-  }
+    HobLength = GetRemainingHobSize (*FoundationHobSize, UsedSize);
+    MmIplBuildResourceHobForAllSystemMemory (FoundationHobList + UsedSize, &HobLength, PlatformHobList, PlatformHobSize, Block, MmProfileDataHob, UnblockRegion, UnblockRegionCount);
```
Comment (jiaxinwu) [inline StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c:939]: MmIplBuildResourceHobForAllSystemMemory (
  FoundationHobList + UsedSize,
  &HobLength,
  PlatformHobList,
  PlatformHobSize,
  Block,
  MmProfileDataHob,
  UnblockRegion,
  UnblockRegionCount
  );

File: StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c [hunk 3]
```diff
@@ -843,23 +916,33 @@ CreateMmFoundationHobList (
...
+    // Collect unblocked memory regions
     //
-    MmIplBuildResourceHobForUnblockedRegion (FoundationHobList + UsedSize, &HobLength);
-  } else {
+    UnblockRegion = CollectUnblockedMemoryRegions (&UnblockRegionCount);
     //
     // All system memory (DRAM) is accessible.
     // When SMM Profile is enabled:
     //   * Access to regions reported from MmPlatformHobProducerLib do not require logging.
     //   * Access to other system memory requires logging.
```
Comment (jiaxinwu) [inline StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c:936]: Correct the comment

File: StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c [hunk 4]
```diff
@@ -843,23 +916,33 @@ CreateMmFoundationHobList (
...
+  //
+  // Build resource HOB for unblocked region
+  //
   HobLength = GetRemainingHobSize (*FoundationHobSize, UsedSize);
-  if (PcdGetBool (PcdCpuSmmRestrictedMemoryAccess)) {
+  MmIplBuildResourceHobForUnblockedRegion (FoundationHobList + UsedSize, &HobLength);
+  UsedSize += HobLength;
+
+  if (!PcdGetBool (PcdCpuSmmRestrictedMemoryAccess)) {
+    UnblockRegion = NULL;
```
Comment (niruiyu) [inline StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c:927]: It's dead code.

File: StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c (hunk 3 above)
Comment (niruiyu) [inline StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c:936]: correct comments.

File: StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c [hunk 5]
```diff
@@ -843,23 +916,33 @@ CreateMmFoundationHobList (
...
     //
-    // Only unblocked memory regions are accessible
+    // Collect unblocked memory regions
     //
-    MmIplBuildResourceHobForUnblockedRegion (FoundationHobList + UsedSize, &HobLength);
-  } else {
+    UnblockRegion = CollectUnblockedMemoryRegions (&UnblockRegionCount);
```
Comment (niruiyu) [inline StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c:931]: The existing code pattern is CollectxxxRegions is called() from MmIplBuildResourceHobForAllSystemMemory.
Your code pattern is NOT like the existing one. Can you follow the exsiting pattern?

Comment (jiaxinwu) [reply StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c:931]: Ray, existing code follow the code pattern that all excluded mem should be passed by the parameter

File: StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c (hunk 4 above)
Comment (hongbin123) [reply StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c:927]: Will remove in next commit

File: StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c (hunk 3 above)
Comment (hongbin123) [reply StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c:936]: Yes, Will correct it.

Comment (hongbin123) [reply StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c:936]: Will correct it in next commit

File: StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c (hunk 2 above)
Comment (hongbin123) [reply StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c]: Updated in new commit

File: StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c (hunk 5 above)
Comment (hongbin123) [reply StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c]: As talked, moved CollectUnblockedMemoryRegions() to MmIplBuildResourceHobForAllSystemMemory()

**Review by niruiyu — APPROVED** [review]

**Review by jiaxinwu — APPROVED** [review]

**Review by niruiyu — APPROVED** [review]

**Pull request merged**